posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,273 @@
+
+``` cpp
+auto iter = TensorIterator();
+iter.add_output(out);
+iter.add_input(a);
+iter.dont_resize_outputs(); // call if out is allocated.
+iter.dont_compute_common_dtype(); // call if inputs/outputs are of a different type.
+iter.build();
+
+auto loop = [&](char **data, const int64_t* strides, int64_t n) {
+    auto * out_data_bytes = data[0];
+    auto * in_data_bytes = data[1];
+    
+    // assume float data type for this example.
+    for (int i = 0; i < n; i++) {
+      (float)(*out_data_bytes) += (float)(*in_data_bytes);
+    }
+    
+    out_data_bytes += strides[0];
+    in_data_bytes += strides[1];
+}
+
+iter.for_each(loop);
+```
+In the above example, the `char** data` gives a pointer to the data within the
+tensor in the same order that you specify when you build the iterator. Note
+that in order to make the implementation agnostic of any particular data type, you
+will always receive the pointer typecast to `char` (think of it as a bunch of bytes).
+
+The second argument is `int64_t* strides` which is an array containing the strides of
+each tensor in the dimension that you're iterating over. We can add this stride to the
+pointer received in order to reach the next element in the tensor. The last argument is
+`int64_t n` which is the size of the dimension being iterated over. 
+
+`for_each` implicitly parallelizes the operation by executing `loop` in parallel
+if the number of iterations is more than the value of `internal::GRAIN_SIZE`, which is a value
+that is determined as the 'right amount' of data to iterate over in order to gain a significant
+speedup using multi-threaded execution. If you want to explicitly specify that your
+operation _must_ run in serial, then use the `serial_for_each` loop.
+
+### Using kernels for iterations
+
+Frequently we want to create a kernel that applies a simple point-wise function onto entire tensors.
+`TensorIterator`
+provides various such generic kernels that can be used for iterating over the elements
+of a tensor without having to worry about the stride, data type of the operands or details
+of the parallelism.
+
+For example, say we want to build a function that performs the point-wise addition
+of two tensors and stores the result in a third tensor, we can use the `cpu_kernel`
+function. Note that in this example we assume a tensor of `float` but you can
+use the `AT_DISPATCH_ALL_TYPES_AND2` macro.
+``` cpp
+TensorIterator iter;
+iter.add_input(a_tensor);
+iter.add_input(b_tensor);
+iter.add_output(c_tensor);
+iter.build();
+cpu_kernel(iter, [] (float a, float b) -> float {
+  return a + b;
+});
+```
+Writing the kernel in this way ensures that the value returned by the lambda passed to
+`cpu_kernel` will populate the corresponding place in the target output tensor.
+
+### Setting tensor iteration dimensions
+
+The value of the strides will determine which dimension of the tensor you will iterate over.
+`TensorIterator` performs optimizations to make sure that at least
```

Comment:
The value of first `sizes`, then `strides`.

Reply:
Changed.